File: src/core/dispatch/state.ts
/**
 * Symphony state-machine vocabulary (SPEC §7.1).
 *
 * Separates "what the user thinks the run is doing" (`RunState`) from
 * "what the runner is doing right now" (`RunPhase`). The first answers
 * "is this task work I can still claim?", the second answers "if it's
 * running, where in the pipeline is it stuck?".
 *
 * The dispatcher emits the corresponding ledger event types so the UI
 * can render real-time per-task progress in the VerificationTimeline
 * (Stage 8 wire-up) and the reconciler can detect stalls
 * (`runReconciler` below).
 *
 * Naming aligns with `reference/symphony/SPEC.md §7.1, §16.5` so the
 * symphony spec is directly applicable to Apohara — same vocabulary,
 * same lifecycle, different host language.
 *
 * G5.B.1 fills the gaps the audit flagged PARCIAL (audit symphony §3):
 *   - explicit `RUN_STATES` closed list,
 *   - `isClaimable(state)` predicate (scheduler input filter),
 *   - `canTransition(from, to)` claim-DAG guard (catches race-driven
 *     stale callers that try `released → running` without re-claim),
 *   - `freshClaimToken()` RFC 4122 v4 (race-free release contract —
 *     orchestrator stores the token alongside the claim and only
 *     accepts release attempts that present the same token),
 *   - `phaseImpliesSuccess` (success ≠ done distinction needed by
 *     continuation chains, see retry-semantics.ts in G5.B.8).
 *
 * G5.B.3 / G5.B.9 / G5.B.10 then layer the secondary concerns
 *   (`BlockedReason`, `CarefulMode`, `TeammateIdle`) on top of the
 *   primary claim states without breaking the vocabulary.
 */

import { randomUUID } from "node:crypto";

export type RunState =
	| "unclaimed"
	| "claimed"
	| "running"
	| "retry_queued"
	| "released";

/** Closed enumeration for callers wanting to iterate every legal claim
 * state (UI legends, migrations, audits). Keep this in sync with the
 * `RunState` union — `canTransition` and `isClaimable` assume the set
 * is complete. */
export const RUN_STATES = [
	"unclaimed",
	"claimed",
	"running",
	"retry_queued",
	"released",
] as const satisfies readonly RunState[];

export type RunPhase =
	| "preparing_workspace"
	| "building_prompt"
	| "launching_agent_process"
	| "initializing_session"
	| "streaming_turn"
	| "finishing"
	| "succeeded"
	| "failed"
	| "timed_out"
	| "stalled"
	| "canceled_by_reconciliation";

/** Per-phase ledger event payload — small and stable. */
export interface PhaseEventPayload {
	taskId: string;
	phase: RunPhase;
	at: number;
	/** Free-form context. Kept lossy on purpose; the orchestration DB
	 * carries the structured per-task state. */
	detail?: string;
}

/** Sentinel value for `RunPhase` events that imply the task ended. */
export const TERMINAL_PHASES: ReadonlySet<RunPhase> = new Set([
	"succeeded",
	"failed",
	"timed_out",
	"stalled",
	"canceled_by_reconciliation",
]);

export function isTerminalPhase(p: RunPhase): boolean {
	return TERMINAL_PHASES.has(p);
}

/**
 * Distinguishes the single "happy ending" phase from the 4 unhappy
 * ones. The continuation pattern (symphony §10.3, see G5.B.8) treats
 * `succeeded` as "this turn ended cleanly — maybe schedule a follow-up
 * if the parent intent still wants more work"; the failure flavours
 * each map to different retry semantics (transient / stall / canceled
 * → see retry-semantics.ts).
 */
export function phaseImpliesSuccess(p: RunPhase): boolean {
	return p === "succeeded";
}

/** Scheduler input filter — picks rows the next tick can pull. */
export function isClaimable(s: RunState): boolean {
	return s === "unclaimed" || s === "released";
}

/**
 * Legal `RunState` transitions per symphony §7.1.
 *
 *   unclaimed   → claimed        (worker picked it up)
 *   claimed     → running        (worker started executing)
 *   claimed     → released       (worker died before starting; reaper)
 *   running     → retry_queued   (transient failure, will retry)
 *   running     → released       (terminal — succeeded or failed)
 *   retry_queued→ claimed        (re-pickup after back-off)
 *   retry_queued→ released       (max retries reached)
 *   released    → unclaimed      (reaper pushes it back into the pool)
 *
 * Any other from/to is REJECTED — callers should not flip `released`
 * straight to `running` (must re-claim), and must not jump from
 * `unclaimed` to `running` (must record the intermediate `claimed` so
 * the audit trail names the responsible worker).
 */
const ALLOWED_TRANSITIONS: ReadonlyMap<RunState, ReadonlySet<RunState>> = new Map([
	["unclaimed", new Set<RunState>(["claimed"])],
	["claimed", new Set<RunState>(["running", "released"])],
	["running", new Set<RunState>(["retry_queued", "released"])],
	["retry_queued", new Set<RunState>(["claimed", "released"])],
	["released", new Set<RunState>(["unclaimed"])],
]);

export function canTransition(from: RunState, to: RunState): boolean {
	return ALLOWED_TRANSITIONS.get(from)?.has(to) ?? false;
}

/**
 * Race-free release contract — RFC 4122 v4. The orchestrator stores
 * this token alongside the claim row at `unclaimed → claimed` and
 * rejects any later `running → released` that doesn't present the same
 * token. Prevents a stale or out-of-band caller from "releasing"
 * someone else's claim (the symphony Elixir code uses ets-table guards
 * for the same purpose; TS we use the bun:sqlite `WHERE token = ?`
 * conditional UPDATE).
 */
export function freshClaimToken(): string {
	return randomUUID();
}

// ---------------------------------------------------------------------
// G5.B.3 — Blocked as primary state (audit symphony §10 PARCIAL → COMPLETO)
// ---------------------------------------------------------------------

/**
 * Closed enumeration of the reasons a task ends up in `blocked` state.
 * The scheduler treats `blocked` as a primary state (separate from
 * `retry_queued` / `released`) so the UI can render "Needs Operator"
 * cards independently from the retry pool.
 *
 *   approval_required           — agent asked to approve a tool call
 *   user_input_required         — agent asked a free-form question
 *   mcp_elicitation             — an MCP server elicited additional
 *                                 spec from the user mid-tool
 *   stalled_after_input_request — input requested but no answer came,
 *                                 and the agent timed out waiting
 *   provider_rejected           — provider hard-refused (TOS, rate
 *                                 limit, auth)
 */
export type BlockedReason =
	| "approval_required"
	| "user_input_required"
	| "mcp_elicitation"
	| "stalled_after_input_request"
	| "provider_rejected";

export interface BlockedSnapshot {
	reason: BlockedReason;
	/** Epoch ms when the block started. Drives `reconciler.PASS_BLOCKED_AGING`. */
	since: number;
	/** Free-form detail (tool name, prompt label, MCP server id). */
	detail?: string;
}

/**
 * Heuristic classifier — takes an event payload coming out of the
 * provider stream and decides if it should park the task in `blocked`
 * state, with which reason.
 *
 * Bias: false negatives > false positives. We'd rather let a run
 * proceed than mis-label a normal completion as `approval_required`
 * and freeze the orchestration queue indefinitely. The classifier
 * therefore returns `null` for any event that isn't an explicit
 * approval / input / elicitation / stall-after-input / rejection
 * signal.
 *
 * Shape mirrors the symphony approach (label keyword search) without
 * coupling to a specific CLI's event vocabulary. Provider drivers
 * adapt their events into this shape before classification.
 */
export interface BlockingEvent {
	kind:
		| "permission_request"
		| "user_input_required"
		| "elicitation"
		| "provider_rejected"
		| "stall"
		| "tool_call_start"
		| "tool_call_end";
	label?: string;
	reason?: string;
	/** When kind === "stall", an earlier timestamp at which we asked
	 * the user for input. Together they trigger
	 * `stalled_after_input_request`. */
	priorInputRequestAt?: number;
}

export function classifyBlocked(ev: BlockingEvent): BlockedSnapshot | null {
	const now = Date.now();
	switch (ev.kind) {
		case "permission_request":
			return { reason: "approval_required", since: now, detail: ev.label };
		case "user_input_required":
			return { reason: "user_input_required", since: now, detail: ev.label };
		case "elicitation":
			return { reason: "mcp_elicitation", since: now, detail: ev.label };
		case "provider_rejected":
			return { reason: "provider_rejected", since: now, detail: ev.reason };
		case "stall":
			if (ev.priorInputRequestAt !== undefined) {
				return {
					reason: "stalled_after_input_request",
					since: now,
					detail: `priorInputAt=${ev.priorInputRequestAt}`,
				};
			}
			return null;
		default:
			return null;
	}
}

// ---------------------------------------------------------------------
// G7.5.A.5 — RunTransition: unified state-mutation payload
// ---------------------------------------------------------------------

/**
 * A state-machine transition that the dispatcher applies as a single
 * unit. The `state` field is `RunState` for the primary claim states
 * plus the secondary `"blocked"` tag (G5.B.3) for tasks parked awaiting
 * operator input. When `state === "blocked"`, the `blockedReason` and
 * `blockedSince` fields MUST be present so the reconciler's blocked-
 * aging pass can escalate stuck inputs.
 *
 * The shape mirrors the optional fields on `BlockedInstruction` in
 * `reconciler.ts` (`blockedSince`, `blockedReason`) so a transition
 * applied to disk doesn't need a separate translation step.
 *
 * G7.5.A.5: introduced when wiring the `classifyBlocked` classifier
 * through the protocol event handler — provider events with
 * `kind: "blocked"` now flow `ProtocolEvent` → `BlockingEvent` →
 * `BlockedSnapshot` → `RunTransition` and the resulting transition
 * carries the specific `BlockedReason` for retry decisioning.
 */
export interface RunTransition {
	state: RunState | "blocked";
	/** Set when `state === "blocked"`. */
	blockedReason?: BlockedReason;
	/** Epoch ms of the block start (set when `state === "blocked"`). */
	blockedSince?: number;
	/** Optional provenance string. */
	detail?: string;
}

File: crates/apohara-dispatch/src/state.rs
//! Run state machine types.
//!
//! Direct port of `src/core/dispatch/state.ts` (TS legacy). The TS module
//! distinguishes:
//!   * `RunState` — the *claim lifecycle* (unclaimed → claimed → running
//!     → retry_queued | released). Bounded, race-free; the scheduler
//!     uses this to decide what's pickable.
//!   * `RunPhase` — *what the runner is doing right now* inside a
//!     `running` claim. Lossy, free-form-ish; the UI uses this to render
//!     real-time per-task progress.
//!
//! Both serialize as snake_case strings so the wire format stays
//! identical during Phase 1 double-maintenance (TS and Rust both feed
//! the same orchestration DB).
//!
//! See `src/core/dispatch/state.ts` for the full design rationale —
//! this module mirrors it so the symphony SPEC.md §7.1 vocabulary
//! survives the port.

use serde::{Deserialize, Serialize};

/// Claim lifecycle states. Closed set; iterate via [`RUN_STATES`].
///
/// TS source: `type RunState = "unclaimed" | "claimed" | "running"
/// | "retry_queued" | "released"`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RunState {
    Unclaimed,
    Claimed,
    Running,
    RetryQueued,
    Released,
}

/// Closed enumeration for callers wanting to iterate every legal claim
/// state (UI legends, migrations, audits). Mirrors TS `RUN_STATES`.
pub const RUN_STATES: [RunState; 5] = [
    RunState::Unclaimed,
    RunState::Claimed,
    RunState::Running,
    RunState::RetryQueued,
    RunState::Released,
];

/// What the runner is doing while it holds a `running` claim.
///
/// TS source: `type RunPhase = "preparing_workspace" | "building_prompt"
/// | "launching_agent_process" | "initializing_session" | "streaming_turn"
/// | "finishing" | "succeeded" | "failed" | "timed_out" | "stalled"
/// | "canceled_by_reconciliation"`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RunPhase {
    PreparingWorkspace,
    BuildingPrompt,
    LaunchingAgentProcess,
    InitializingSession,
    StreamingTurn,
    Finishing,
    Succeeded,
    Failed,
    TimedOut,
    Stalled,
    CanceledByReconciliation,
}

impl RunPhase {
    /// Mirrors TS `isTerminalPhase` — phases that imply the task ended,
    /// regardless of outcome.
    #[inline]
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            RunPhase::Succeeded
                | RunPhase::Failed
                | RunPhase::TimedOut
                | RunPhase::Stalled
                | RunPhase::CanceledByReconciliation
        )
    }
}

/// Distinguishes the single "happy ending" phase from the 4 unhappy ones.
/// Mirrors TS `phaseImpliesSuccess`. Used by the continuation pattern
/// (symphony §10.3) to decide whether to schedule follow-up work.
#[inline]
pub fn phase_implies_success(p: RunPhase) -> bool {
    matches!(p, RunPhase::Succeeded)
}

/// Scheduler input filter — picks rows the next tick can pull.
/// Mirrors TS `isClaimable`.
#[inline]
pub fn is_claimable(s: RunState) -> bool {
    matches!(s, RunState::Unclaimed | RunState::Released)
}

/// Legal `RunState` transitions per symphony §7.1. Mirrors TS
/// `ALLOWED_TRANSITIONS` / `canTransition`.
///
/// Any other from/to is REJECTED — callers should not flip `Released`
/// straight to `Running` (must re-claim), and must not jump from
/// `Unclaimed` to `Running` (must record the intermediate `Claimed` so
/// the audit trail names the responsible worker).
pub fn can_transition(from: RunState, to: RunState) -> bool {
    use RunState::*;
    matches!(
        (from, to),
        (Unclaimed, Claimed)
            | (Claimed, Running)
            | (Claimed, Released)
            | (Running, RetryQueued)
            | (Running, Released)
            | (RetryQueued, Claimed)
            | (RetryQueued, Released)
            | (Released, Unclaimed)
    )
}

/// Race-free release contract — RFC 4122 v4. The orchestrator stores
/// this token alongside the claim row at `unclaimed → claimed` and
/// rejects any later `running → released` that doesn't present the
/// same token. Mirrors TS `freshClaimToken` (which used Node's
/// `crypto.randomUUID`).
pub fn fresh_claim_token() -> String {
    uuid::Uuid::new_v4().to_string()
}

/// Closed enumeration of the reasons a task ends up in `blocked` state.
///
/// TS source: `type BlockedReason = "approval_required"
/// | "user_input_required" | "mcp_elicitation"
/// | "stalled_after_input_request" | "provider_rejected"`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BlockedReason {
    /// Agent asked to approve a tool call.
    ApprovalRequired,
    /// Agent asked a free-form question.
    UserInputRequired,
    /// An MCP server elicited additional spec from the user mid-tool.
    McpElicitation,
    /// Input was requested but no answer came; the agent timed out
    /// waiting.
    StalledAfterInputRequest,
    /// Provider hard-refused (TOS, rate limit, auth).
    ProviderRejected,
}

/// The discriminator field on [`RunTransition`].
///
/// TS source: `state: RunState | "blocked"` — a flat string-union.
/// We represent it as a Rust enum where one branch wraps the primary
/// claim state and the other is the secondary `"blocked"` tag, with a
/// custom serde rule that flattens both into a single string on the
/// wire (so `{"state":"claimed"}` and `{"state":"blocked"}` both
/// round-trip cleanly through the same struct).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransitionState {
    /// One of the primary claim-lifecycle states.
    Run(RunState),
    /// The secondary `"blocked"` tag — task is parked awaiting operator
    /// input. When this variant is set, [`RunTransition::blocked_reason`]
    /// and [`RunTransition::blocked_since`] MUST be populated so the
    /// reconciler's blocked-aging pass can escalate stuck inputs.
    Blocked,
}

impl Serialize for TransitionState {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            TransitionState::Run(s) => s.serialize(serializer),
            TransitionState::Blocked => serializer.serialize_str("blocked"),
        }
    }
}

impl<'de> Deserialize<'de> for TransitionState {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let s: String = String::deserialize(deserializer)?;
        match s.as_str() {
            "blocked" => Ok(TransitionState::Blocked),
            "unclaimed" => Ok(TransitionState::Run(RunState::Unclaimed)),
            "claimed" => Ok(TransitionState::Run(RunState::Claimed)),
            "running" => Ok(TransitionState::Run(RunState::Running)),
            "retry_queued" => Ok(TransitionState::Run(RunState::RetryQueued)),
            "released" => Ok(TransitionState::Run(RunState::Released)),
            other => Err(serde::de::Error::unknown_variant(
                other,
                &[
                    "unclaimed",
                    "claimed",
                    "running",
                    "retry_queued",
                    "released",
                    "blocked",
                ],
            )),
        }
    }
}

/// A state-machine transition that the dispatcher applies as a single
/// unit.
///
/// Direct port of TS `RunTransition` (G7.5.A.5). The `state` field is
/// `RunState | "blocked"` in TS — see [`TransitionState`].
///
/// When `state == TransitionState::Blocked`, the [`Self::blocked_reason`]
/// and [`Self::blocked_since`] fields MUST be present so the
/// reconciler's blocked-aging pass can escalate stuck inputs (mirrored
/// from `reconciler.ts` PASS_BLOCKED_AGING).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunTransition {
    pub state: TransitionState,
    /// Set when `state == Blocked`.
    #[serde(rename = "blockedReason", skip_serializing_if = "Option::is_none")]
    pub blocked_reason: Option<BlockedReason>,
    /// Epoch ms of the block start (set when `state == Blocked`).
    #[serde(rename = "blockedSince", skip_serializing_if = "Option::is_none")]
    pub blocked_since: Option<u64>,
    /// Optional provenance string.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}
